conda/core/subdir_data.py

```diff
@@ -98,6 +98,22 @@ def __call__(cls, channel, repodata_fn=REPODATA_FN):
         return subdir_data_instance
 
 
+class LazyPackageRecords(UserList):
+    """
+    Lazily convert dicts to PackageRecord.
+    """
+
+    def __getitem__(self, i):
+        if isinstance(i, slice):
+            return self.__class__(self.data[i])
+        else:
+            record = self.data[i]
+            if not isinstance(record, PackageRecord):
+                record = PackageRecord(**record)
+                self.data[i] = record
+            return record
```

Comment:
conda-build wants to iterate over `_package_records`  so this list keeps things compatible, and could simplify the rest of this PR (but the iter_records() with looser guarantees is nice)